File: ml2d.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
# Morris-Lecar model on a 2D lattice
# FvW 03/2018

import os
import numpy as np
import matplotlib.pyplot as plt
import cv2


def ml2d(N, T, t0, dt, s, D, gL, VL, gCa, VCa, gK, VK, C, I0,
         V0, V1, V2, V3, V4, phi, stim, blocks):
    C1 = 1.0/C
    # initialize Morris-Lecar system
    v, w = V0*np.ones((N,N)), 0.015*np.ones((N,N))
    dv, dw = np.zeros((N,N)), np.zeros((N,N))
    s_sqrt_dt = s*np.sqrt(dt)
    X = np.zeros((T,N,N))
    # stimulation protocol
    I = np.zeros((t0+T,N,N))
    for st in stim:
        t_on, t_off = st[0]
        x0, x1 = st[1]
        y0, y1 = st[2]
        I[t0+t_on:t0+t_off, x0:x1, y0:y1] = I0

    # steady-state functions
    m_inf = lambda v: 0.5*(1 + np.tanh((v-V1)/V2))
    w_inf = lambda v: 0.5*(1 + np.tanh((v-V3)/V4))
    lambda_w = lambda v: phi*np.cosh((v-V3)/(2*V4))

    # iterate
    for t in range(1, t0+T):
        if (t%100 == 0): print(f"    t = {t:d}/{t0+T:d}\r", end="")
        # ML equations
        dv = 1/C*(-gL*(v-VL)-gCa*m_inf(v)*(v-VCa)-gK*w*(v-VK)+I[t,:,:])+D*L(v)
        dw = lambda_w(v) * (w_inf(v) - w)
        # Ito stochastic integration
        v += (dv*dt + s_sqrt_dt*np.random.randn(N,N))
        w += (dw*dt)
        # dead block(s):
        for bl in blocks:
            v[bl[0][0]:bl[0][1], bl[1][0]:bl[1][1]] = 0.0
            w[bl[0][0]:bl[0][1], bl[1][0]:bl[1][1]] = 0.0
        if (t >= t0):
            X[t-t0,:,:] = v
    print("\n")
    return X


def animate_video(fname, x):
    # BW
    y = 255 * (x-x.min()) / (x.max()-x.min())
    # BW inverted
    #y = 255 * ( 1 - (x-x.min()) / (x.max()-x.min()) )
    y = y.astype(np.uint8)
    nt, nx, ny = x.shape
    #print(f"nt = {nt:d}, nx = {nx:d}, ny = {ny:d}")
    # write video using opencv
    frate = 60
    out = cv2.VideoWriter(fname, \
                          cv2.VideoWriter_fourcc(*'mp4v'), \
                          frate, (nx,ny))
    print("[+] Animate:")
    for i in range(0,nt):
        print(f"\ti = {i:d}/{nt:d}\r", end="")
        img = np.ones((nx, ny, 3), dtype=np.uint8)
        for j in range(3): img[:,:,j] = y[i,::-1,:]
        out.write(img)
    out.release()
    print("")


def L(x):
    # Laplace operator
    # periodic boundary conditions
    xU = np.roll(x, shift=-1, axis=0)
    xD = np.roll(x, shift=1, axis=0)
    xL = np.roll(x, shift=-1, axis=1)
    xR = np.roll(x, shift=1, axis=1)
    Lx = xU + xD + xL + xR - 4*x
    # non-periodic boundary conditions
    Lx[0,:] = 0.0
    Lx[-1,:] = 0.0
    Lx[:,0] = 0.0
    Lx[:,-1] = 0.0
    return Lx


def main():
    print("Morris-Lecar (ML) lattice model\n")
    # Example: D=0.8, gCa=5, gK=6, C=5, s=2.0
    N = 128
    T = 20000
    t0 = 1000
    dt = 0.01
    s = 2
    D = 0.8
    # Morris-Lecar parameters
    gL = 2.0
    VL = -60.0
    gCa = 5.0 # 4.0
    VCa = 120.0
    gK = 6.0 # 8.0
    VK = -84 # -80.0 #-84.0
    C = 5.0 # 20.0
    V0 = -60.0
    V1 = -1.2
    V2 = 18.0
    V3 = 12.0 # 2.0
    V4 = 17.4 # 30.0
    phi = 1/15 # 0.04
    I = 90.0 # 110
    print("[+] Lattice size N: ", N)
    print("[+] Time steps T: ", T)
    print("[+] Warm-up steps t0: ", t0)
    print("[+] Integration time step dt: ", dt)
    print("[+] Noise intensity s: ", s)
    print("[+] Diffusion coefficient D: ", D)
    print("[+] ML parameter gL: ", gL)
    print("[+] ML parameter VL: ", VL)
    print("[+] ML parameter gCa: ", gCa)
    print("[+] ML parameter VCa: ", VCa)
    print("[+] ML parameter gK: ", gK)
    print("[+] ML parameter VK: ", VK)
    print("[+] ML parameter C: ", C)
    print("[+] ML parameter V0: ", V0)
    print("[+] ML parameter V1: ", V1)
    print("[+] ML parameter V2: ", V2)
    print("[+] ML parameter V3: ", V3)
    print("[+] ML parameter V4: ", V4)
    print("[+] ML parameter phi: ", phi)
    print("[+] Stimulation current I: ", I)

    # stimulation current configuration
    stim = [ [[50,250], [0,5], [0,5]],
             [[6200,6700], [25,30], [0,15]] ]
    #stim = [ [[50,250], [0,5], [0,5]],
    #         [[6200,6700], [25,30], [0,15]] ]
    #stim = []

    # dead blocks, array of elementy [[x0,x1], [y0,y1]]
    blocks = [ [[0,15], [5,10]] ]
    #blocks = []

    # run simulation
    data = ml2d(N, T, t0, dt, s, D, gL, VL, gCa, VCa, gK, VK, C, I,
                V0, V1, V2, V3, V4, phi, stim, blocks)
    print("[+] data dimensions: ", data.shape)

    # plot mean voltage
    m = np.mean(np.reshape(data, (T,N*N)), axis=1)
    plt.figure(figsize=(12,4))
    plt.plot(m, "-k")
    plt.tight_layout()
    plt.show()

    # save data
    #fname1 = f"ml2d_I_{I:.2f}_s_{s:.2f}_D_{D:.2f}.npy"
    #np.save(fname1, data)
    #print("[+] Data saved as: ", fname1)

    # video
    fname2 = f"ml2d_I_{I:.2f}_s_{s:.2f}_D_{D:.2f}.mp4"
    animate_video(fname2, data)
    print("[+] Data saved as: ", fname2)


if __name__ == "__main__":
    os.system("clear")
    main()

File: ml2d.jl
#!/usr/local/bin/julia
# last tested Julia version: 1.6.1
# Morris-Lecar model on a 2D lattice
# FvW 03/2018

using NPZ
using PyCall
using PyPlot
using Statistics
using VideoIO
@pyimport matplotlib.animation as anim

function ml2d(N, T, t0, dt, s, D, gL, VL, gCa, VCa, gK, VK, C, I0,
			  V0, V1, V2, V3, V4, phi, stim, blocks)
    C1 = 1.0/C
    # initialize Morris-Lecar system
    v = V0*ones(Float64,N,N)
    w = 0.015*ones(Float64,N,N)
    dv = zeros(Float64,N,N)
    dw = zeros(Float64,N,N)
    s_sqrt_dt = s*sqrt(dt)
    X = zeros(Float64,T,N,N)
    # stimulation protocol
    I = zeros(Float64,t0+T,N,N)
    for st in stim
        t_on, t_off = st[1]
        x0, x1 = st[2]
        y0, y1 = st[3]
        I[t0+t_on:t0+t_off, x0:x1, y0:y1] .= I0
    end

    # iterate
    for t in range(1, stop=t0+T, step=1)
        (t%100 == 0) && print("    t = ", t, "/", t0+T, "\r")
        # ML equations
        dv = C1.*(-gL.*(v.-VL) -gCa.*0.5.*(1.0 .+ tanh.((v.-V1)/V2)).*(v.-VCa)
                 -gK.*w.*(v.-VK) .+ I[t,:,:]) .+ D.*L(v)
        dw = phi*cosh.((v.-V3)/(2*V4)) .* (0.5*(1.0 .+ tanh.((v.-V3)/V4))-w)
        # Ito stochastic integration
        v += (dv*dt + s_sqrt_dt*randn(N,N))
        w += (dw*dt)
        # dead block(s):
        for bl in blocks
            v[bl[1][1]:bl[1][2], bl[2][1]:bl[2][2]] .= 0.0
            w[bl[1][1]:bl[1][2], bl[2][1]:bl[2][2]] .= 0.0
        end
        (t > t0) && (X[t-t0,:,:] = v)
    end
    println("\n")
    return X
end

function animate_pyplot(fname, data)
    """
    Animate 3D array as .mp4 using PyPlot, save as `fname`
    array dimensions:
        1: time
        2, 3: space
    """
    nt, nx, ny = size(data)
    vmin = minimum(data)
    vmax = maximum(data)
    # setup animation image
    fig = figure(figsize=(6,6))
    axis("off")
    t = imshow(data[1,:,:], origin="lower", cmap=ColorMap("gray"),
			   vmin=vmin, vmax=vmax)
    tight_layout()
    # frame generator
    println("[+] animate")
    function animate(i)
        (i%100 == 0) && print("    t = ", i, "/", nt, "\r")
        t.set_data(data[i+1,:,:])
    end
    # create animation
    ani = anim.FuncAnimation(fig, animate, frames=nt, interval=10)
    println("\n")
    # save animation
    ani[:save](fname, bitrate=-1,
               extra_args=["-vcodec", "libx264", "-pix_fmt", "yuv420p"])
    show()
end

function animate_video(fname, data)
    """
    Animate 3D array as .mp4 using VideoIO, save as `fname`
    array dimensions:
        1: time
        2, 3: space
    """
    # BW
    y = UInt8.(round.(255*(data .- minimum(data))/(maximum(data)-minimum(data))))
    # BW inverted
    #y = UInt8.(round.(255 .- 255*(data .- minimum(data))/(maximum(data)-minimum(data))))
    encoder_options = (color_range=2, crf=0, preset="medium")
    framerate=30
    T = size(data,1)
    open_video_out(fname, y[1,end:-1:1,:], framerate=framerate,
                   encoder_options=encoder_options) do writer
        for i in range(2,stop=T,step=1)
            write(writer, y[i,end:-1:1,:])
        end
    end
end

function L(x)
    # Laplace operator
    # periodic boundary conditions
    xU = circshift(x, [-1 0])
    xD = circshift(x, [1 0])
    xL = circshift(x, [0 -1])
    xR = circshift(x, [0 1])
    Lx = xU + xD + xL + xR - 4x
    # non-periodic boundary conditions
    Lx[1,:] .= 0.0
    Lx[end,:] .= 0.0
    Lx[:,1] .= 0.0
    Lx[:,end] .= 0.0
    return Lx
end

function main()
    println("Morris-Lecar (ML) lattice model\n")
    N = 128
    T = 15000
    t0 = 1000
	dt = 0.01
	s = 2.0
	D = 0.8
	# Morris-Lecar parameters
    gL = 2.0
    VL = -60.0
    gCa = 5.5 # 4.0
    VCa = 120.0
    gK = 8.0
    VK = -84.0
    C = 5.0 # 20.0
    I = 120.0
    V0 = -60.0
    V1 = -1.2
    V2 = 18.0
    V3 = 12.0 # 2.0
    V4 = 17.4 # 30.0
    phi = 1.0/15.0 # 0.04
    println("[+] Lattice size N: ", N)
    println("[+] Time steps T: ", T)
    println("[+] Warm-up steps t0: ", t0)
	println("[+] Integration time step dt: ", dt)
	println("[+] Noise intensity s: ", s)
	println("[+] Diffusion coefficient D: ", D)
    println("[+] ML parameter gL: ", gL)
    println("[+] ML parameter VL: ", VL)
    println("[+] ML parameter gCa: ", gCa)
    println("[+] ML parameter VCa: ", VCa)
    println("[+] ML parameter gK: ", gK)
    println("[+] ML parameter VK: ", VK)
    println("[+] ML parameter C: ", C)
    println("[+] ML parameter V0: ", V0)
    println("[+] ML parameter V1: ", V1)
    println("[+] ML parameter V2: ", V2)
    println("[+] ML parameter V3: ", V3)
    println("[+] ML parameter V4: ", V4)
    println("[+] ML parameter phi: ", phi)
	println("[+] Stimulation current I: ", I)

    # stim protocol, array of elements [[t0,t1], [x0,x1], [y0,y1]]
    # double spiral 1
    #stim = [ [[50,250], [1,5], [1,5]],
    #         [[5800,6200], [20,25], [1,35]] ]
    # double spiral 2
    #stim = [ [[50,250], [1,5], [1,5]],
    #         [[5800,6200], [18,23], [5,26]] ]
    # spiral wave 1
    #stim = [ [[50,250], [1,5], [1,5]],
    #         [[6300,6800], [30,35], [5,20]] ]
    # spiral wave 2
    stim = [ [[50,250], [1,5], [1,5]],
             [[6400,6900], [30,35], [5,20]] ]
    #stim = []

    # dead blocks, array of elementy [[x0,x1], [y0,y1]]
    blocks = [ [[1,15], [5,10]] ]
    #blocks = []

    # run simulation
    data = ml2d(N, T, t0, dt, s, D, gL, VL, gCa, VCa, gK, VK, C, I,
				V0, V1, V2, V3, V4, phi, stim, blocks)
    println("[+] Data dimensions: ", size(data))

    # plot mean voltage
    m = mean(reshape(data, (T,N*N)), dims=2)
    plot(m, "-k"); show()

    # save data
    I_str = rpad(I, 4, '0') # stim. current amplitude as 4-char string
    s_str = rpad(s, 4, '0') # noise as 4-char string
    D_str = rpad(D, 4, '0') # diffusion coefficient as 4-char string
    #fname1 = string("ml2d_I_", I_str, "_s_", s_str, "_D_", D_str, ".npy")
    #npzwrite(fname1, data)
    #println("[+] Data saved as: ", fname1)

    # video
    fname2 = string("ml2d_I_", I_str, "_s_", s_str, "_D_", D_str, ".mp4")
    #animate_pyplot(fname2, data) # slow
    animate_video(fname2, data) # fast
    println("[+] Data saved as: ", fname2)
end

main()
